Thy rework ir

Soulthym opened the pull request:
New graph structure for the IR

One bug is left: when using the builder pattern, nodes get duplicated into multiple parents if a call is delegated to a sub-node. I suspect it's due to the automatic unlinking from the old parent (it also removes it from the old parent's children) getting triggered while the parent is still unset.

Soulthym committed:
rework Operation: base structure

Soulthym committed:
fix ir2 module structure

Soulthym committed:
implement Sub and Mul

Leo-Besancon committed:
Add helpers to insert nodes for each operation

Soulthym committed:
make Operands dyn Value
implement Enf, Call, Fold, For, If, Vector, Matrix, Boundary, Parameter

Soulthym committed:
Generic visitor

Soulthym committed:
expose roots and graph to visitor

Soulthym committed:
inlining with generic visitor

Soulthym committed:
blanket migration for degree and trace
added Eq support to operation
few bugfixes

Soulthym committed:
implement post order visitor

Soulthym committed:
fix missing visit of stored nodes

Soulthym committed:
Rename VisitOrder::DepthFirst to Manual
implement real depth first
remove external stack in post-order visit
abstract Visit trait with auto implem for VisitContext implems

Soulthym committed:
added Eq, Clone and Debug support to Value and Operations

Soulthym committed:
convert MirValue to the new structure

Leo-Besancon committed:
Add unrolling and lowering

Soulthym committed:
ir2: new graph structure based on enums

Soulthym committed:
Split graph module, and several improvements to the graph datastructure

Leo-Besancon commented on ir/src/ir2/graph/graph.rs:
```diff
@@ -0,0 +1,178 @@
+            self
+        );
+        let x = children.borrow()[0].clone();
+        x
```
A more Rusty way to write it would be:something like:
```suggestion
        self.get_children()
            .borrow()
            .first()
            .expect("first() called on node without children: {:?}", self)
            .clone()
```
(same for last below)

Soulthym replied:
Noted and fixed for the next commit on .first()

However, the let x seems necessary for it to compile due to lifetimes being dropped in the wrong order, the fix is also a compiler suggestion

The error was:
```
`children` dropped here while still borrowed... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `Ref<'_, Vec<Link<NodeTypes>>>`
```

Leo-Besancon commented on ir/src/ir2/graph/graph.rs:
```diff
@@ -0,0 +1,178 @@
+            self
+        );
+        let x = children.borrow()[0].clone();
+        x
```
A more Rusty way to write it would be:

```suggestion
        self.get_children()
            .borrow()
            .first()
            .expect(format!("first() called on node without children: {:?}", self).as_str())
            .clone()
```
(same for the last below)

Leo-Besancon commented on ir/src/ir2/graph/nodes.rs:
```diff
@@ -0,0 +1,236 @@
+impl Child for Link<NodeTypes> {
+    fn get_parent(&self) -> BackLink<NodeTypes> {
+        match self.borrow().deref() {
+            NodeTypes::Graph(_) => panic!("get_parent called on graph"),
```
I assume the goal of this NodeTypes is to represent nodes that don't have a parent.
IMO it's clearer to have a NodeTypes::Root that only reference one "Node", and then a Graph is basically a `Vec<Root>` (or a `Graph { boundary_roots: Vec<Root>, integrity_roots: Vec<Root> }`).

Would that be possible? Or does this require too much change / is not possible because of some edge case?

The main reason not to do it I think is if this duplicates code to handle adding roots to a graph or similar

Soulthym replied:
It is possible, but that currently means duplicating the code almost verbatim for the graph builder pattern (modulo what's generic and what's behind the enum variant)

I was thinking of splitting NodeType(s) into multiple categories, but havent yet found a way to make that generic enough to avoid big code duplication. If I find a way to, I'll make it a separate one with just the graph. For now, this has been the easiest to implement/extend upon.

Leo-Besancon commented on ir/src/ir2/graph/nodes.rs:
```diff
@@ -0,0 +1,236 @@
+            NodeTypes::I32(ref mut leaf) => leaf.swap_parent(parent),
+            NodeTypes::Add(ref mut add) => add.swap_parent(parent),
+            NodeTypes::Function(ref mut function) => function.swap_parent(parent),
+            _ => panic!("set_parent called on non-node"),
```
maybe "... on a NodeType that does not implement Child" is clearer?

Leo-Besancon commented on ir/src/ir2/graph/nodes.rs:
```diff
@@ -0,0 +1,236 @@
+        self.node.get_children().borrow().deref()[1].clone()
+    }
+    pub fn body(&self) -> Link<NodeTypes> {
+        self.node.get_children().borrow().deref()[2].clone()
```
Keep in mind that index access can panic.
I don't think adding error handling is necessary, but maybe we can ensure that Function can only be created through Function::new() and without direct Object instantiation. Here, the field `node: Node` is private so it's fine, but on Add below it's not for example.

So nothing to change in this case, but just commenting to share :)

Leo-Besancon commented on ir/src/ir2/graph/nodes.rs:
```diff
@@ -0,0 +1,236 @@
+            }
+        }
+        if not_a_node {
+            panic!("add_child called on non-node: {:?}", self);
```
weird naming there, I think having an helper function `is_node` / `can_have_children` on NodeTypes is clearer. Then, once you check, you can have "unreachable".
and change the message to "on a NodeTypes that does not impl Parent" maybe

Soulthym replied:
Good call, thats probably a way to customize the node's behaviour in traits with default implementations too, via such helpers.

Leo-Besancon commented on ir/src/ir2/graph/nodes.rs:
```diff
@@ -0,0 +1,236 @@
+use std::ops::{Deref, DerefMut};
+
+#[derive(Clone, Eq, PartialEq)]
+pub enum NodeTypes {
```
I think the convention is to use singular NodeType to have, when you reference variants: `NodeType::Function(_)`

Leo-Besancon commented on ir/src/ir2/graph/nodes.rs:
```diff
@@ -0,0 +1,236 @@
+pub enum NodeTypes {
+    Graph(Graph),
+    Node(Node),
+    I32(Leaf<i32>),
```
Would this be Value(Leaf<MirValue>) basically?
And you'd have something like Variable(Leaf<MirVariable>) along side it as the two Leaf types?

Soulthym replied:
Yes pretty much, the Leaf is just used as a special case where it has no children. It's meant as a way to represent raw values, types, etc. Elements that cannot have sub-nodes essentially. I found it easier to think of it as a link to an arbitrary struct that should not link to other nodes (other than their parent).

Variable probably can be one too, but I need to think about edgecases more before committing to that answer. As a first guess, I agree.

The I32 naming is a relique from my testing, I'll rename to Value.

Leo-Besancon reviewed (commented):
Great work! Left a few comments mainly about the code, but will probably need some more time and dig deeper into it to really understand the usage and what can be improved when using it :)

Soulthym committed:
fix builder pattern duplication and split+rename NodeType
rename NodeTypes to NodeType
split NodeType into LeafNode, RootNode, MiddleNode
removed debugging

Soulthym committed:
replace asserts by expects
new_i32 replaced by generic new_value
LeafNode::I32 renamed to Value
implemented Felt Leaf node
improved conversions
removed debugging

Soulthym committed:
Scope with unique insertions
removed MiddleNode::Node
renamed new_body -> new_scope
implement Default

Soulthym committed:
IsNode derive macro

Soulthym committed:
fix module name
